pkgs/stdenv/generic/setup.sh

```diff
@@ -334,36 +334,49 @@ appendToVar() {
     fi
 }
 
-# Accumulate into `flagsArray` the flags from the named variables.
+# Accumulate flags from the named variables $2+ into the indexed array $1.
 #
-# If __structuredAttrs, the variables are all treated as arrays
-# and simply concatenated onto `flagsArray`.
-#
-# If not __structuredAttrs, then:
-#   * Each variable is treated as a string, and split on whitespace;
-#   * except variables whose names end in "Array", which are treated
-#     as arrays.
-_accumFlagsArray() {
-    local name
-    if [ -n "$__structuredAttrs" ]; then
-        for name in "$@"; do
-            local -n nameref="$name"
-            flagsArray+=( ${nameref+"${nameref[@]}"} )
-        done
-    else
-        for name in "$@"; do
+# Arrays are simply concatenated, strings are split on whitespace.
+concatTo() {
+    local -n targetref="$1"; shift
+    local name type
+    for name in "$@"; do
+        if type=$(declare -p "$name" 2> /dev/null); then
             local -n nameref="$name"
-            case "$name" in
-                *Array)
-                    # shellcheck disable=SC2206
-                    flagsArray+=( ${nameref+"${nameref[@]}"} ) ;;
+            case "${type#* }" in
+                # Deals with associative arrays by only taking the values,
+                # because that's the most sensible thing to do. Letting it
+                # fallthrough to *) would result in only the first value being
+                # taken. It's expected to be called with either indexed arrays
+                # or strings, but maybe somebody comes up with a clever use-case
+                # for an associative array later.
+                -A*|-a*)
+                    targetref+=( "${nameref[@]}" ) ;;
                 *)
                     # shellcheck disable=SC2206
-                    flagsArray+=( ${nameref-} ) ;;
+                    targetref+=( ${nameref-} ) ;;
             esac
-        done
-    fi
+        fi
+    done
+}
 
+# Place separator $1 between each element of a list and return the string.
+# For arrays each item is separated, for strings whitespace is replaced.
```

Comment:
"each item is separated" - I really fail to read this one; I think/hope this just means that the array is joined on the separator, original items kept intact, like so:

```console
$ a=("foo bar" "biba boba")
$ intersperse ; a
foo bar;biba boba
```

Reply:
I think we should existing phrasing as much as we can, and then clarify the behaviour for the lists:

- https://nix.dev/manual/nix/2.23/language/builtins.html#builtins-concatStringsSep
- https://docs.python.org/3/library/stdtypes.html#str.join

E.g.

> `concatStringsSep separator array`
>
> Concatenate an array of strings with a separator between each element.
> The second input may be a string in which case it is split on whitespaces:
>
> ```console
> $ flags="a b c"
> $ concatStringsSep ";" flags
> a;b;c
> ```
>
> ```console
> $ flags=("lorem ipsum" "dolor" "sit amet")
> $ concatStringsSep ";" flags
> lorem ipsum;dolor;sit amet
> ```

Reply:
The actual corresponding Haskell function is `intercalate`, though I can’t exactly recommend that name.